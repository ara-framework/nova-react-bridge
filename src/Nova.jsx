import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { v4 } from 'uuid';

const { window, CustomEvent, document } = global;

class Nova extends Component {
  constructor(props) {
    super(props);
    this.uuid = v4();
    this.placeholder = React.createRef();
  }

  componentDidMount() {
    if (window !== 'undefined') {
      const id = this.placeholder.current.getAttribute('data-hypernova-id');
      const { name } = this.props;
      const customEvent = new CustomEvent('NovaMount', { detail: { id, name } });
      document.dispatchEvent(customEvent);
    }
  }

  componentDidUpdate() {
    if (window !== 'undefined') {
      const id = this.placeholder.current.getAttribute('data-hypernova-id');
      const { name } = this.props;
      const customEvent = new CustomEvent('NovaMount', { detail: { id, name } });
      document.dispatchEvent(customEvent);
    }
  }

  render() {
    const { name, data } = this.props;
    return (
      <>
        <div
          ref={this.placeholder}
          data-hypernova-key={name}
          data-hypernova-id={this.uuid}
          dangerouslySetInnerHTML={{ __html: '' }} // eslint-disable-line react/no-danger
        />
        <script
          type="application/json"
          data-hypernova-key={name}
          data-hypernova-id={this.uuid}
          dangerouslySetInnerHTML={{ __html: `<!--${JSON.stringify(data)}-->` }} // eslint-disable-line react/no-danger
        />
      </>
    );
  }
}

Nova.propTypes = {
  name: PropTypes.string.isRequired,
  data: PropTypes.shape({}).isRequired,
};

export default Nova;
