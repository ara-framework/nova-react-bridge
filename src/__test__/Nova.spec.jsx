import React from 'react';
import { mount } from 'enzyme';

import Nova from '../Nova';

jest.mock('uuid', () => ({
  v4: () => 'uuid',
}));

describe('Nova', () => {
  test('render placeholder with script tag correctly', () => {
    const data = { title: 'Ara' };

    const wrapper = mount(
      <Nova name="Example" data={data} />,
    );

    expect(wrapper.html()).toMatchSnapshot();
  });

  test('render placeholder with default data correctly', () => {
    const wrapper = mount(
      <Nova name="Example" />,
    );

    expect(wrapper.html()).toMatchSnapshot();
  });

  test('emit NovaMount when component is mount', async () => {
    const promise = new Promise((resolve) => {
      document.addEventListener('NovaMount', resolve);
    });

    mount(
      <Nova name="Example" />,
    );

    const event = await Promise.resolve(promise);

    expect(event.detail).toMatchSnapshot();
  });
});
