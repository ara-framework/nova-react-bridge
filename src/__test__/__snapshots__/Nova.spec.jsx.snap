// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nova emit NovaMount when component is mount 1`] = `
Object {
  "id": "uuid",
  "name": "Example",
}
`;

exports[`Nova render placeholder with default data correctly 1`] = `"<div data-hypernova-key=\\"Example\\" data-hypernova-id=\\"uuid\\"></div><script type=\\"application/json\\" data-hypernova-key=\\"Example\\" data-hypernova-id=\\"uuid\\"><!--{}--></script>"`;

exports[`Nova render placeholder with script tag correctly 1`] = `"<div data-hypernova-key=\\"Example\\" data-hypernova-id=\\"uuid\\"></div><script type=\\"application/json\\" data-hypernova-key=\\"Example\\" data-hypernova-id=\\"uuid\\"><!--{\\"title\\":\\"Ara\\"}--></script>"`;
